scene/2d/shape_cast_2d.cpp

```diff
@@ -0,0 +1,467 @@
+				} else {
+					exclude.erase(Object::cast_to<CollisionObject2D>(get_parent())->get_rid());
+				}
+			}
+		} break;
+		case NOTIFICATION_EXIT_TREE: {
+			if (enabled) {
+				set_physics_process_internal(false);
+			}
+		} break;
+
+		case NOTIFICATION_DRAW: {
+#ifdef TOOLS_ENABLED
+			ERR_FAIL_COND(!is_inside_tree());
+			if (!Engine::get_singleton()->is_editor_hint() && !get_tree()->is_debugging_collisions_hint()) {
+				break;
+			}
+			if (shape.is_null()) {
+				break;
+			}
+			Color draw_col = get_tree()->get_debug_collisions_color();
+			if (!enabled) {
+				float g = draw_col.get_v();
+				draw_col.r = g;
+				draw_col.g = g;
+				draw_col.b = g;
+			}
+			// Draw continuos chain of shapes along the cast.
+			const int steps = MAX(2, target_position.length() / shape->get_rect().get_size().length() * 4);
+			for (int i = 0; i <= steps; ++i) {
+				Vector2 t = (real_t(i) / steps) * target_position;
+				draw_set_transform(t, 0.0, Size2(1, 1));
+				shape->draw(get_canvas_item(), draw_col);
+			}
+			draw_set_transform(Vector2(), 0.0, Size2(1, 1));
+
+			// Draw an arrow indicating where the ShapeCast is pointing to.
+			if (target_position != Vector2()) {
+				Transform2D xf;
+				xf.rotate(target_position.angle());
+				xf.translate(Vector2(target_position.length(), 0));
+
+				draw_line(Vector2(), target_position, draw_col, 2);
+				Vector<Vector2> pts;
+				float tsize = 8;
+				pts.push_back(xf.xform(Vector2(tsize, 0)));
+				pts.push_back(xf.xform(Vector2(0, Math_SQRT12 * tsize)));
+				pts.push_back(xf.xform(Vector2(0, -Math_SQRT12 * tsize)));
+				Vector<Color> cols;
+				for (int i = 0; i < 3; i++)
+					cols.push_back(draw_col);
+
+				draw_primitive(pts, cols, Vector<Vector2>());
+			}
+#endif
+		} break;
+		case NOTIFICATION_INTERNAL_PHYSICS_PROCESS: {
+			if (!enabled) {
+				break;
+			}
+			_update_shapecast_state();
+		} break;
+	}
+}
+
+void ShapeCast2D::_update_shapecast_state() {
+	result.clear();
+
+	ERR_FAIL_COND_MSG(shape.is_null(), "Invalid shape.");
+
+	Ref<World2D> w2d = get_world_2d();
+	ERR_FAIL_COND(w2d.is_null());
+
+	PhysicsDirectSpaceState2D *dss = PhysicsServer2D::get_singleton()->space_get_direct_state(w2d->get_space());
+	ERR_FAIL_COND(!dss);
+
+	PhysicsDirectSpaceState2D::ShapeParameters params;
+	params.shape_rid = shape_rid;
+	params.transform = get_global_transform();
+	params.motion = get_global_transform().basis_xform(target_position);
+	params.margin = margin;
+	params.exclude = exclude;
+	params.collision_mask = collision_mask;
+	params.collide_with_bodies = collide_with_bodies;
+	params.collide_with_areas = collide_with_areas;
+
+	if (target_position != Vector2()) {
+		dss->cast_motion(params, collision_safe_distance, collision_unsafe_distance);
+	}
+	bool intersected = true;
+	while (intersected && result.size() < max_results) {
+		PhysicsDirectSpaceState2D::ShapeRestInfo info;
+		intersected = dss->rest_info(params, &info);
```

Comment:
I've found `rest_info()` to be the best compromise in 3.x, because:

1. You can detect multiple bodies along the cast given detected bodies get subsequently excluded from intersection:

![image](https://user-images.githubusercontent.com/17108460/141168946-0712b295-9813-4fc5-a619-35cc2b0358a8.png)

2. It does return useful information such as normal.

Reply:
I agree the position and normal would be good to have, but given the current API it just returns collisions in a random order and the positions are not correct so it's not really usable in common cases.

What you're trying to do makes sense. I would just rather wait for making the proper changes in the physics API before exposing the extra information to avoid confusing users, since it's not working as intended for now.

Reply:
What if we keep `rest_info()` when `motion == Vector2()`, and use `intersect_shape()` when `motion != Vector2()`? Would that be a good compromise?

Reply:
Yes, that could be a good compromise as long as it's clear and well documented. It does sound like a better solution than using a separate node.

Reply:
It was all wrong from the start due to porting error, see https://github.com/godotengine/godot/pull/54803#issuecomment-966407480.